compiler/rustc_hir_typeck/src/fn_ctxt/suggestions.rs

```diff
@@ -1687,4 +1687,18 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
             false
         }
     }
+
+    pub(crate) fn is_field_suggestable(&self, field: &ty::FieldDef, hir_id: HirId) -> bool {
+        // The field must be visible in the containing module.
+        field.vis.is_accessible_from(self.tcx.parent_module(hir_id), self.tcx)
+            // The field must not be unstable.
+            && !matches!(
+                self.tcx.eval_stability(field.did, None, rustc_span::DUMMY_SP, None),
+                rustc_middle::middle::stability::EvalResult::Deny { .. }
+            )
+            // If the field is from an external crate it must not be `doc(hidden)`.
+            && (field.did.is_local() || !self.tcx.is_doc_hidden(field.did))
+            // If the field is hygienic it must not originate from a macro expansion.
+            && !self.tcx.def_ident_span(field.did).unwrap().normalize_to_macros_2_0().from_expansion()
```

Comment:
should we check that the span ctxts are the same? Does this affect our ability to suggest hygenic fields from *within* a macro?

Reply:
Have to check, good points.

Reply:
Maybe test something like:

```
#![feature(decl_macro)]

pub macro foo() {
    fn test() {
        struct Hello { world: () }
        let x: Hello = Hello { worl: () };
    }
}

foo!();
```

Reply:
Yep, you're right. It's obvious in hindsight.

Reply:
No worries. We have lots of diagnostics that are subtly wrong due to being overly defensive, resulting them just totally not applying within macros even when they should. Thx for investigating.